### PR TITLE
[Bug Fix]: Deem hash repartition unnecessary when input and output has 1 partition

### PR DESCRIPTION
**Pull request by mustafasrepo**
## Which issue does this PR close?

Closes [#9928](https://github.com/apache/arrow-datafusion/issues/9928).

## Are these changes tested?
Yes

### CODE REVIEW

**Comment by mustafasrepo** [comment]
cc @echai58, this fix should solve the problem in the issue. Feel free to review, if you have time.

**Comment by mustafasrepo** [comment]
@korowa thanks again for the [analysis](https://github.com/apache/arrow-datafusion/issues/9928#issuecomment-2054147818) which helped a lot in understanding the issue and replicating it in datafusion.

**Comment by alamb** [comment]
Thanks @mustafasrepo  and @ozankabak

**Comment by korowa** [comment]
@mustafasrepo good catch on UNION -- I previously suggested that this kind of plan (with partitions > config.target_partitions) is "illegal" in DF.

**Comment by mustafasrepo** [comment]
> @mustafasrepo good catch on UNION -- I previously suggested that this kind of plan (with partitions > config.target_partitions) is "illegal" in DF.

I agree, this behaviour is a bit counter intuitive. However, with current implementation of the `UnionExec` it is hard to make sure `partitions<=config.target_partitions` always `true`. If this behaviour prone to errors. Maybe we can insert `RepartitionExec` on top `UnionExec`s if their output partition number > `config.target_partitions`. By this way, we can guarantee this violation wouldn't propagate to other operators.

File: datafusion/sqllogictest/test_files/joins.slt [hunk 1]
```diff
@@ -3587,3 +3587,104 @@ SELECT 1 FROM join_partitioned_table JOIN (SELECT c1 AS id1 FROM join_partitione
...
+05)--------PlaceholderRowExec
+06)------CoalesceBatchesExec: target_batch_size=2
+07)--------RepartitionExec: partitioning=Hash([c@0], 1), input_partitions=2
+08)----------UnionExec
+09)------------ProjectionExec: expr=[1 as c, 2 as d]
+10)--------------PlaceholderRowExec
+11)------------ProjectionExec: expr=[1 as c, 3 as d]
+12)--------------PlaceholderRowExec
+
+query IIII
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/joins.slt:3676]: I think this query is non determisitic and fails sometimes on CI as it doesn't have an `ORDER BY` and isn't annotated with `rowsort`. Here is a PR to fix that: https://github.com/apache/arrow-datafusion/pull/10120

**Comment by korowa** [comment]
> Maybe we can insert RepartitionExec on top UnionExecs if their output partition number > config.target_partitions. By this way, we can guarantee this violation wouldn't propagate to other operators.

Probably better solution would be planning union inputs execution according to total available partitions -- e.g
```sql
    select l_linenumber as f
    from lineitem
    union all
    select l_orderkey as f
    from lineitem
```
with target_partitions = 4, could plan 2 threads for each ParquetExec (ideally we could also use byte/row statistics and plan according to them -- not only 2-2, but probably 1-3 if there is significant data skew across inputs/files).

And on top of it, when target_partitions is less then number of UNION inputs (e.g. UNION has 10 inputs, target_partitions = 4, and we need at least 1 thread for each input) there could be RepartitionExec.

**Comment by mustafasrepo** [comment]
> > Maybe we can insert RepartitionExec on top UnionExecs if their output partition number > config.target_partitions. By this way, we can guarantee this violation wouldn't propagate to other operators.
> 
> Probably better solution would be planning union inputs execution according to total available partitions -- e.g
> 
> ```sql
>     select l_linenumber as f
>     from lineitem
>     union all
>     select l_orderkey as f
>     from lineitem
> ```
> 
> with target_partitions = 4, could plan 2 threads for each ParquetExec (ideally we could also use byte/row statistics and plan according to them -- not only 2-2, but probably 1-3 if there is significant data skew across inputs/files).
> 
> Currently, with target_partitions = 4, it's planned as 4 threads per ParquetExec, and 8 output partitions for UNION.
> 
> And on top of it, when target_partitions is less then number of UNION inputs (e.g. UNION has 10 inputs, target_partitions = 4, and we need at least 1 thread for each input) there could be RepartitionExec.

That might work. However, this approach cannot solve all cases I guess. For the following query
 ```sql
select * from table
union all
select * from table
union all 
select * from table
 ```sql
when `target_partitions=2`, there is no way to change input partitions to generate 2 partitions after `union`. We would generate at least `3` partitions(assuming each input generates single partition). Hence, your suggestion may not solve all cases.

**Comment by echai58** [comment]
Hi guys, not familiar with datafusion's release process - is there a estimate of when this will be released in a new datafusion version?

**Comment by alamb** [comment]
> Hi guys, not familiar with datafusion's release process - is there a estimate of when this will be released in a new datafusion version?

This should be included in 38.0.0 -- I just filed https://github.com/apache/datafusion/issues/10255 to track that release if you want to watch it  @echai58

**Pull request merged**